python/dgl/node_flow.py

```diff
@@ -0,0 +1,845 @@
+
+    See Also
+    --------
+    dgl.DGLGraph.edges
+    """
+    __slots__ = ['_graph', '_flow']
+
+    def __init__(self, graph, flow):
+        self._graph = graph
+        self._flow = flow
+
+    def __getitem__(self, key):
+        return self._graph._edge_frames[self._flow][key]
+
+    def __setitem__(self, key, val):
+        self._graph._edge_frames[self._flow][key] = val
+
+    def __delitem__(self, key):
+        del self._graph._edge_frames[self._flow][key]
+
+    def __len__(self):
+        return len(self._graph._edge_frames[self._flow])
+
+    def __iter__(self):
+        return iter(self._graph._edge_frames[self._flow])
+
+    def __repr__(self):
+        data = self._graph._edge_frames[self._flow]
+        return repr({key : data[key] for key in data})
+
+def _copy_to_like(arr1, arr2):
+    return F.copy_to(arr1, F.context(arr2))
+
+def _get_frame(frame, names, ids):
+    col_dict = {name: frame[name][_copy_to_like(ids, frame[name])] for name in names}
+    if len(col_dict) == 0:
+        return FrameRef(Frame(num_rows=len(ids)))
+    else:
+        return FrameRef(Frame(col_dict))
+
+
+def _update_frame(frame, names, ids, new_frame):
+    col_dict = {name: new_frame[name] for name in names}
+    if len(col_dict) > 0:
+        frame.update_rows(ids, FrameRef(Frame(col_dict)), inplace=True)
+
+
+class NodeFlow(DGLBaseGraph):
+    """The NodeFlow class stores the sampling results of Neighbor sampling and Layer-wise sampling.
+
+    These sampling algorithms generate graphs with multiple layers. The edges connect the nodes
+    between two layers while there don't exist edges between the nodes in the same layer.
+
+    We store multiple layers of the sampling results in a single graph. We store extra information,
+    such as the node and edge mapping from the NodeFlow graph to the parent graph.
+
+    Parameters
+    ----------
+    parent : DGLGraph
+        The parent graph
+    graph_index : NodeFlowIndex
+        The graph index of the NodeFlow graph.
+    """
+    def __init__(self, parent, graph_idx):
+        super(NodeFlow, self).__init__(graph_idx)
+        self._parent = parent
+        self._node_mapping = graph_idx.node_mapping
+        self._edge_mapping = graph_idx.edge_mapping
+        self._layer_offsets = graph_idx.layers.tonumpy()
+        self._block_offsets = graph_idx.flows.tonumpy()
+        self._node_frames = [FrameRef(Frame(num_rows=self.layer_size(i))) \
+                             for i in range(self.num_layers)]
+        self._edge_frames = [FrameRef(Frame(num_rows=self.block_size(i))) \
+                             for i in range(self.num_blocks)]
+        # registered functions
+        self._message_funcs = [None] * self.num_blocks
+        self._reduce_funcs = [None] * self.num_blocks
+        self._apply_node_funcs = [None] * self.num_blocks
+        self._apply_edge_funcs = [None] * self.num_blocks
+
+    def _get_layer_id(self, layer_id):
```

Comment:
Docstring. "Handle layer_id < 0 case, blahblahblah ..."

Reply:
do we need to write docstring for hidden methods?